hsds/link_sn.py

```diff
@@ -76,31 +62,79 @@ async def GET_Links(request):
 
     await validateAction(app, domain, group_id, username, "read")
 
-    req = getDataNodeUrl(app, group_id)
-    req += "/groups/" + group_id + "/links"
+    kwargs = {"bucket": bucket}
 
-    params = {}
-    if create_order:
-        params["CreateOrder"] = 1
-    if limit is not None:
-        params["Limit"] = str(limit)
-    if marker is not None:
-        params["Marker"] = marker
-    if bucket:
-        params["bucket"] = bucket
-    links_json = await http_get(app, req, params=params)
-    log.debug(f"got links json from dn for group_id: {group_id}")
-    links = links_json["links"]
-
-    # mix in collection key, target and hrefs
-    for link in links:
-        if link["class"] == "H5L_TYPE_HARD":
-            collection_name = getCollectionForId(link["id"])
-            link["collection"] = collection_name
-            target_uri = "/" + collection_name + "/" + link["id"]
-            link["target"] = getHref(request, target_uri)
-        link_uri = "/groups/" + group_id + "/links/" + link["title"]
-        link["href"] = getHref(request, link_uri)
+    if "follow_links" in params and params["follow_links"]:
+        follow_links = True
+    else:
+        follow_links = False
+
+    if "pattern" in params and params["pattern"]:
+        pattern = params["pattern"]
+        try:
+            globmatch("abc", pattern)
+        except ValueError:
+            msg = f"invlaid pattern: {pattern} for  link matching"
+            log.warn(msg)
+            raise HTTPBadRequest(reason=msg)
+        log.debug(f"using pattern: {pattern} for GET_Links")
+        kwargs["pattern"] = pattern
+    else:
+        pattern = None
+
+    if follow_links:
+        # Use DomainCrawler to fetch links from multiple objects.
+        # set the follow_links and bucket params
+        kwargs["follow_links"] = True
+        crawler_kwargs = {"action": "get_link", "raise_error": True, "params": kwargs}
+        items = [group_id, ]
+        crawler = DomainCrawler(app, items, **crawler_kwargs)
+
+        # will raise exception on NotFound, etc.
+        await crawler.crawl()
+
+        msg = f"DomainCrawler returned: {len(crawler._obj_dict)} objects"
+        log.info(msg)
+        links = crawler._obj_dict
+        if pattern:
+            for grp_id in links.keys():
+                grp_links = links[grp_id]
+                ret_links = []
+                for link in grp_links:
+                    title = link["title"]
+                    if globmatch(title, pattern):
```

Comment:
Is this use of globmatch redundant, since `domain_crawl.sn/get_links` -> `getLinks` ->`GET_Links (DN)` already does the pattern matching?